File: 1408-find-the-smallest-divisor-given-a-threshold/find-the-smallest-divisor-given-a-threshold.js
/**
 * @param {number[]} nums
 * @param {number} threshold
 * @return {number}
 */
var smallestDivisor = function (nums, threshold) {

    let max = Math.max(...nums);

    let low = 1;
    let high = max;
    let ans = -1;

    while (low <= high) {
        let mid = Math.floor((low + high) / 2);
        
        if (calAns(mid, nums) <= threshold) {
            ans = mid;
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }


    function calAns(mid, nums) {
        let sum = 0;
        for (let j = 0; j < nums.length; j++) {
            sum = sum + Math.ceil(nums[j] / mid);
        }
        return sum;
    }

    return ans;

};

File: 1408-find-the-smallest-divisor-given-a-threshold/find-the-smallest-divisor-given-a-threshold.java
class Solution {
  int check(int divisor, int[] nums) {
    int sum = 0;
    for (int num : nums) {
      sum += num / divisor + (num % divisor == 0 ? 0 : 1);
    }

    return sum;
  }

  public int smallestDivisor(int[] nums, int threshold) {
    int max = nums[0];

    for (int num : nums) {
      max = Math.max(max, num);
    }

    int low = 1;
    int high = max;
    int res = max;
    while (low <= high) {
      int mid = (low + high) / 2;

      if (check(mid, nums) <= threshold) {
        res = mid;
        high = mid - 1;
      } else {
        low = mid + 1;
      }

    }

    return res;
  }
}